### DISCUSSION TITLE
watch() vs useController() for read-only fields

### DISCUSSION THREAD

**Discussion by msageryd** [react-hook-form/react-hook-form#7557] [Q&A]
I have quite complex forms with multiple pages. Some fields may occur more than once on different pages. In those cases there is often one input field and one or more read-only texts. The read-only texts are not input fields, only simple text renderings.

The read-only fields are packaged as specific components. I call them PreviewField.

In the PreviewField component I could either just subscribe to the field value via `watch()`, or i could mimic the internals of a fully fledged inputField by using `useController`.

**watch** seems like a more lightweight approach for the use case. But I do wonder if there is any substantial difference between watch and useController for this use case.

**Comment by Moshyfawn** [comment]
Hm, I've never considered using `useController` for read-only fields. I don't think it's gonna work, tho. `useController` doesn't subscribe to the field's value, per say. See [this CSB](https://codesandbox.io/s/controller-vs-watch-example-84e4b).
Also, you might wanna consider [`useWatch`](https://react-hook-form.com/api/usewatch/#main) hook instead of the `watch` form method to isolate re-rendering at the component level.

**Reply by msageryd** [reply]
I'm gonna use `useWatch`, since it seems most appropriate.

But I'm curious. Why wouldn't `useController` work? 
My fields gets their values from useController.

```
  const {
    field: {onChange, onBlur, value, ref},
    fieldState: {invalid, isTouched, isDirty},
    formState: {touchedFields, dirtyFields},
  } = useController({
    name,
    control,
  });
```

If I update a fieldValue from elsewhere, the inputField will immediately show the new value, without everytning else rerendering. This looks very much like some kind of subscription to me.

**Reply by msageryd** [reply]
Hmm. I see in your CSB that your fields does not update via useController. But mine do. I'm now investigating what the difference could be.

**Comment by msageryd** [comment]
This is really weird. I don't know why my `useControll` acts like a subscription and your example fields does not.
- One difference is that I'm using React-Native.
- I also have a slightly different approach. I use `useController` instead of `Controller`

Here is a simplified example of one of my components.

```
const TextInputField = ({name}) => {
  const {control} = useFormApi();

  const {
    field: {onChange, value, ref},
  } = useController({
    name,
    control,
  });

  return (
    <>
      <TextInput
        onChangeText={onChange}
        value={value}
        ref={ref}
      />
      <RenderCount />
    </>
  );
```

Here is my "root" view for the form. Nothing fancy. Some of the methods I send with the context are from `useForm`.
```
    <View style={{flex: 1}}>
      <FormApiContext.Provider
        value={{
          control,
          setValue,
          handleSubmit,
          formState,
          formDef,
          fieldSet,
          uiState,
          dispatchCondition,
          debug: {
            debugState,
            toggleShowPaths,
            toggleShowData,
            toggleShowI18n,
            toggleShowConditions,
            toggleShowErrors,
            toggleShowRenderCount,
          },
        }}
      >
        <RegisterFields />
        <FormBuilder />

      </FormApiContext.Provider>
    </View>
```

`RegisterFields` is quite simple. It iterates through all fieldDefinitions and registers each fields. Validator functions are attached as well.

`FormBuilder` is where the field components gets rendered like this: `<TextField name={'firstName'}/>`

**Reply by msageryd** [reply]
@Moshyfawn 
I have put together a simple example in Expo Snack. Snack is kind of a Code Sandbox for React Native. I'm rendering two TextInputs with the same fieldname. If you write in one of them, the other gets updated.

Why doesn't your example behave like this?

https://snack.expo.dev/@msageryd/rhf-subscription-test

**Reply by Moshyfawn** [reply]
If you look at [my CSB example](https://codesandbox.io/s/controller-vs-watch-example-84e4b), you can see that I register my inputs with `register` form method, which "collects" the input's value via `ref` which **doesn't update any state** _(which makes the input **uncontrolled**, as I don't **"feed" the input it's own value** back)_.

Now, using `useController` in my `<PreviewFieldController />`, the only prop I use from it is `field.value`. I don't pass `field.onChange` into anything _(mostly because I render a `<p />` with the field's value only lol :P)_. Nor do I "feed" the field's value into anything _(just displaying it)_.

[In your Snack](https://snack.expo.dev/@msageryd/rhf-subscription-test) you **pass** both **`field.value`** and **`field.onChange` into the `<TextField />`'s `value` and `onChangeText` props accordingly**. Which actually makes the component **controlled** _(hence the Controller / useController name)_. So, both the first and the second inputs update the same `fistName` field

**Reply by Moshyfawn** [reply]
_*Note on `register` form method._
Actually, I haven't looked into RHF source code for a while. It might be that `register` collects the input's value from `onChange`, not `ref`. But if you look at props it returns...
```
onChange: ChangeHandler;
onBlur: ChangeHandler;
ref: RefCallBack;
name: InternalFieldName;
```
... there's no value returned. Which makes the input uncontrolled.

**Reply by msageryd** [reply]
hmm. more questions than answers in my head right now.

Q1. It seems to me that you pass both value and onChange to the input in both your cases.
`{...field} is tha same as {onChange, value, ref, onBlur, name}`, isn't it?
` <input id={name} {...field} />`

`register` returns onChange, onBlur value, ref and name, i.e. `{...register(name)}` equals `{onChange, value, onBlur, name}`
` <input id={name} {...register(name)} />`

Did I misunderstand soemthing here?

Q2. I still don't understand why one is controlled and the other is not. Is there a difference between the `onChange` returned by `register` respective `useController`?

Q3. If I would go the uncontrolled way (maybe for perfomance?) I would still need to feed my fields with values and handle the onChange. The difference seems to be that onChange should call setValue and the value has to be subscribed to via useWatch. Is this correct? If correct, will there be any perfomance gains left? 

I suspect that I could as well keep using the controlled concept. My forms often have 10-30 fields. Some of them will have complicated relations between them, but otherwise the forms are quite straight forward.

**Reply by msageryd** [reply]
Oh, you are right. There is no value returned. I think I understand a bit better now.
One last question (thank you for your patience with me =).

The relations between my fields can lead to FieldA getting a value due to some condition in FieldB. 
FieldA might not have been rendered yet, if it sits on another page (sub-form).

Hence I need to register all fields so I can have RHF store the value of FieldA until it's actually rendered.

Have I understood the mechanics correctly? Or should I solve my case in another way?

**Comment by Moshyfawn** [comment]
Fair enough! :P
So, yeah, `register` **doesn't return value**, which defines it as **uncontrolled**, since the "source of truth" for it's value is in the DOM _(as opposed to being in React state)_. Using the controller API _(`<Controller />` component or `useController` hook)_, [you can still create an uncontrolled field](https://codesandbox.io/s/uncontrolled-controller-example-se9br?file=/src/App.tsx) _(by not passing the `field.value` prop into the input's value prop)_, but controller basically has **it's own state** "in React", so there's DOM input value and controller `field.value`. That 's the reason [for my CSB ](https://codesandbox.io/s/controller-vs-watch-example-84e4b) behave the way it behaves with the input being registered by `register` form method and `<Controller />` "not subscribed" to the input's value.

In regards to fields relations.. that's a pretty broad topic with no clear answer, I'm afraid. At least, I haven't made up my mind about having shipped an app with a couple of wizard, sub, nested forms using RHF.

There's [a Wizard Form / Funnel section in RHF docs](https://react-hook-form.com/advanced-usage#WizardFormFunnel), which describes storing currently displayed fields in an external (out of RHF) state on form submission _(each step is a separate form)_. Which implies that all the "calculation" of a dependent field will be happening on the other form component mount. So, you don't register all the fields, but pass the necessary values on form submission. Which can totally be used on opening the sub-form via `getValues` form method all the same.
Registering all the necessary fields, no matter if displayed or not, is complete valid as well. There're a couple different ways to go about it, from what I can tell. But I think I can only suggest you try the approach you're thinking of to see if there's any caveat.

Me personally, I did the separate forms with external state on submit thing just because it was in the RHF docs and I was on a time crunch for release. I didn't like it all that much, since I'm using validation schemas [with resolvers](https://github.com/react-hook-form/resolvers) instead of the field level validation, so had to split things, duplicate things, think where / when / how to store stuff from form to form..

And finally, you are always welcome! I don't have all the time in the world to go through with all of the discussion at once, because job and life and all, but I'm all for stuff like this, that might even surface some of the issues, API decisions to consider, etc. I'm happy to help adopt RHF basically is what I'm saying lol :P
